doc: Add introduction of using Windows Guest OS (#3389)
projectacrn/acrn-hypervisor: doc/tutorials/using_windows_as_uos.rst

.. _using_windows_as_uos:

Using Windows as Guest OS on ACRN
#################################
This tutorial describes how to launch Windows as the Guest OS on the ACRN hypervisor.

Hardware setup
**************
The following Intel Kaby Lake NUCs are verified:

.. csv-table::
   :header: "Platform Model", "Bios Version", "Download Link"

   "NUC7i7DNHE", "DNKBLi7v.86A.0052.2018.0808.1344", "`link <https://downloadcenter.intel.com/download/28886?v=t>`__"
   "NUC7i5DNHE", "DNKBLi5v.86A.0060.2018.1220.1536", "`link <https://downloadcenter.intel.com/download/28885?v=t>`__"

ACRN Service OS Setup
*********************
You may refer to the steps in :ref:`getting-started-apl-nuc` for

> Intel NUC to setup ACRN on KBL NUC. After following the steps in that guide,

Reviewer: Intel NUC to setup ACRN on KBL NUC.

Change to:
Intel NUC to set up ACRN on the KBL NUC

         # scp <acrn.efi from your host> /boot/EFI/acrn/
         # scp <acrn-dm from your host> /usr/bin/
         # chmod +x /usr/bin/acrn-dm && sync

Build Service OS Kernel
=======================
   #. Follow the steps described at :ref:`getting-started-building` to set up the build environment
   #. Follow the steps below to build ACRN kernel::

      $ WORKDIR=`pwd`;
      $ JOBS=`nproc`
      $ git clone -b master https://github.com/projectacrn/acrn-kernel.git
      $ cd acrn-kernel && mkdir -p ${WORKDIR}/{build,build-rootfs}
      $ cp kernel_config_uefi_sos ${WORKDIR}/build/.config
      $ make olddefconfig O=${WORKDIR}/build && make -j${JOBS} O=${WORKDIR}/build
      $ make modules_install INSTALL_MOD_PATH=${WORKDIR}/build-rootfs O=${WORKDIR}/build -j${JOBS}

Update Kernel on KBL NUC
========================
   #. Copy new kernel image (bzImage) and it's modules to target machine::

      # scp <your host>:$WORKDIR/build/arch/x86/boot/bzImage /boot/bzImage
      # scp -r <your host>:$WORKDIR/build-rootfs/lib/modules/* /lib/modules/
      # cp /boot/loader/entries/acrn.conf  /boot/loader/entries/acrngt.conf

   #. Modify ``acrngt.conf`` to the content as given below:

> 

Reviewer: After you left-align the numbered list, please move the code block info to the left as well:
#. Modify ``acrngt.conf`` to the content as given below:
    .. code-block:: none
       title The ACRNGT Service OS
       linux /bzImage

      .. code-block:: none

          title The ACRNGT Service OS
          linux /bzImage
          options console=tty0 console=ttyS0 root=/dev/sda3 rw rootwait ignore_loglevel no_timer_check consoleblank=0 i915.nuclear_pageflip=1 i915.avail_planes_per_pipe=0x010101 i915.domain_plane_owners=0x011100001111 i915.enable_gvt=1 i915.enable_conformance_check=0 i915.enable_guc=0 hvlog=2M@0x1FE00000

>       .. note:: You should change ``/dev/sda3`` to your file system partition.

Reviewer: Please left-align notes:
.. note:: You should change ``/dev/sda3`` to your file system partition.

   #. ``reboot`` Service OS and select ``The ACRNGT Service OS`` from the boot menu to apply the updates
      of ACRN kernel and hypervisor

Create Windows 10 Image
=======================
Now we need to create a Windows 10 Image which need two steps:

>    #. Re-generate an ISO which with virtio-win drivers and Windows graphics drivers

Reviewer: Left-align numbered lists:
#. Re-generated an ISO . . .

#. Delete ``C:\Dev\Temp\wim\windows10-17763-107-LTSC\sources\boot.wim`` and
   ``C:\Dev\Temp\wim\windows10-17763-107-LTSC\sources\install.wim``

#. Copy ``C:\WIM\boot.wim`` and ``C:\WIM\install.wim`` to ``C:\Dev\Temp\wim\windows10-17763-107-LTSC\sources``

#. Download and unzip `cdrtools-3.01.a23-bootcd.ru-mkisofs.7z
   <http://reboot.pro/index.php?app=core&module=attach&section=attach&attach_id=15214>`_ to a folder, in my case,
   it is unzipped to ``C:\Dev\Temp\wim\cdrtools-3.01.a23-bootcd.ru-mkisofs``

#. Download the `mkisofs_both_legacy_and_uefi.bat
   <https://raw.githubusercontent.com/projectacrn/acrn-hypervisor/master/doc/scripts/mkisofs_both_legacy_and_uefi.bat>`_
   to a folder in your Windows PC

   .. note:: You will need to change these parameters to your case: ``inputdir``, ``outputiso``,
      ``mkisofs.exe path``

#. The ISO will be generated in ``outputiso`` where you've specified in the script above

Create Raw Disk
---------------
Run these commands on Service OS::

   # swupd bundle-add kvm-host
   # mkdir /root/img
   # cd /root/img
   # qemu-img create -f raw win10-ltsc-virtio.img 30G

Install Windows 10
------------------

> Currently ACRNGT OVMF GOP driver is not ready, a special VGA version is used to install Windows 10

Author: > IMPORTANT: Do you know in what release the driver will be ready??

@chejianj Do you have any idea about this question ?

> Currently ACRNGT OVMF GOP driver is not ready, a special VGA version is used to install Windows 10

Reviewer 2: The plan is that OVMF ACRNGT GOP driver will be ready at the end of Q3

> Currently ACRNGT OVMF GOP driver is not ready, a special VGA version is used to install Windows 10

Reviewer: Thank you @lirui34 and @chejianj!

                , "GOP & VNC remote display",        "Working"
    "GVT-g",      "GVT-g without local display",     "Working with 3D benchmark"
           ,      "GVT-g  with local display",       "Working with 3D benchmark"
    "Tools",      "WinDbg",                          "Working"
    "Test cases", "Install Windows 10 from scratch", "OK"
                , "Windows reboot",                  "OK"
                , "Windows shutdown",                "OK"
    "Built-in Apps", "Microsoft Edge",               "OK"
                   , "Maps",                         "OK"
                   , "Microsoft Store",              "OK"
                   , "3D Viewer",                    "OK"

Known Limitations
*****************
* Cursor is not visible with GVG-g local display.
* Windows graphic driver version must be ``dch_win64_25.20.100.6444.exe``;
  the latest version ``dch_win64_25.20.100.6577.exe`` has problem with ``KVMGT`` and can’t be installed correctly.

Device configurations of acrn-dm command line
*********************************************
* *-s 3,ahci,hd:/root/img/win10.img*:
  This is the hard disk to install windows 10 onto. Please remember the slot ID 3 here.

>   Please make sure it points to your win10 img path.

Reviewer: ```suggestion
  Make sure that the slot ID 3 points to your win10 img path.
```

* *-s 4,virtio-net,tap0*:
  This is for the network virtualization.

* *-s 5,fbuf,tcp=0.0.0.0:5900,w=800,h=600*:
  This will open a port 5900 on SOS which can be connected to via vncviewer.

* *-s 6,virtio-input,/dev/input/event4*:
  This is to passthrough the mouse/keyboard to Windows via virtio.
  Please change ``event4`` accordingly. You can use the following command to check the event node on your SOS::

   <To get the input event of mouse>
   # cat /proc/bus/input/devices | grep mouse

* *-s 7,ahci,cd:/root/img/Windows.iso*:
  This is the IOS image used to install Windows 10. It appears as a cdrom device.

>   Please remember the slot ID 7 here and make sure it points to your win10 ISO path

Reviewer: ```suggestion
Make sure that the slot ID 7 points to your win10 ISO path.
```

* *-s 8,ahci,cd:/root/img/virtio-win-0.1.141.iso*: This is another cdrom device

>   to install virtio windows driver later. Please make sure it points to your VirtIO ISO path.

Reviewer: ```suggestion
  to install the virtio Windows driver later. Make sure it points to your VirtIO ISO path.
```

* *--ovmf /root/bios/OVMF.fd*:

>   Please make sure it points to your OVMF binary. Please make sure it points to your OVMF binary path.

Reviewer: ```suggestion
  Make sure it points to your OVMF binary path.
```
